graph/formats/rdf/rdf.go

```diff
@@ -0,0 +1,386 @@
+// interface.
+func (s *Statement) ReversedEdge() graph.Edge {
+	r := *s
+	r.Subject, r.Object = r.Object, r.Subject
+	return &r
+}
+
+// ReversedLine returns a graph.Edge with the subject and object of the
+// statement swapped. It is implemented to satisfy the graph.Line
+// interface, but may not be a meaningful operation on an RDF statement
+// without a change of the predicate.
+func (s *Statement) ReversedLine() graph.Line {
+	r := *s
+	r.Subject, r.Object = r.Object, r.Subject
+	return &r
+}
+
+// ParseNQuad parses the statement and returns the corresponding Statement.
+// All Term UID fields are zero on return.
+func ParseNQuad(statement string) (*Statement, error) {
+	s, err := parse([]rune(statement))
+	if err != nil {
+		return nil, err
+	}
+	return &s, err
+}
+
+// Decoder is an RDF stream decoder. Statements returned by calls to the
+// Unmarshal method have their Terms' UID fields set so that unique terms
+// will have unique IDs and so can be used directly in a graph.Multi, or
+// in a graph.Graph if all predicate terms are identical. IDs created by
+// the decoder all exist within a single namespace and so Terms can be
+// uniquely identified by their UID.
+type Decoder struct {
+	scanner *bufio.Scanner
+
+	strings store
+	ids     map[string]int64
+}
+
+// NewDecoder returns a new Decoder that takes input from r.
+func NewDecoder(r io.Reader) *Decoder {
+	return &Decoder{
+		scanner: bufio.NewScanner(r),
+		strings: make(store),
+		ids:     make(map[string]int64),
+	}
+}
+
+// Reset resets the decoder to use the provided io.Reader, retaining
+// the existing Term ID mapping.
+func (dec *Decoder) Reset(r io.Reader) {
+	dec.scanner = bufio.NewScanner(r)
+}
+
+// Unmarshal returns the next statement from the input stream.
+func (dec *Decoder) Unmarshal() (*Statement, error) {
+	for dec.scanner.Scan() {
+		data := bytes.TrimSpace(dec.scanner.Bytes())
+		if len(data) != 0 && data[0] != '#' {
+			s, err := ParseNQuad(string(data))
+			if err != nil {
+				return nil, fmt.Errorf("rdf: failed to parse %q: %w", data, err)
+			}
+			if s == nil {
+				continue
+			}
+			s.Subject.Value = dec.strings.intern(s.Subject.Value)
+			s.Predicate.Value = dec.strings.intern(s.Predicate.Value)
+			s.Object.Value = dec.strings.intern(s.Object.Value)
+			s.Subject.UID = dec.idFor(s.Subject.Value)
+			s.Object.UID = dec.idFor(s.Object.Value)
+			s.Predicate.UID = dec.idFor(s.Predicate.Value)
+			if s.Label.Value != "" {
+				s.Label.Value = dec.strings.intern(s.Label.Value)
+				s.Label.UID = dec.idFor(s.Label.Value)
+			}
+			return s, nil
+		}
+	}
+	dec.strings = nil
+	err := dec.scanner.Err()
+	if err != nil {
+		return nil, err
+	}
+	return nil, io.EOF
+}
+
+func (dec *Decoder) idFor(s string) int64 {
+	id, ok := dec.ids[s]
+	if ok {
+		return id
+	}
+	id = int64(len(dec.ids))
+	dec.ids[s] = id
+	return id
+}
+
+// Terms returns the mapping between terms and graph node IDs constructed
+// during decoding the RDF statement stream.
```

Comment:
This and the comment on `Unmarshal` should note that the IDs are keyed on the unescaped quoted node text as it appears in the original statement stream (reflected in the contents of the `Term.Value` field and no folding is performed.